pkg/fanotify/fanotify_linux.go

```diff
@@ -0,0 +1,440 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package fanotify
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/errdefs"
+	"github.com/sirupsen/logrus"
+
+	"github.com/containerd/console"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/stargz-snapshotter/analyzer/fanotify"
+	runtimespec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/spf13/cobra"
+)
+
+type Context struct {
+	Enable             bool
+	Opts               []oci.SpecOpts
+	fanotifier         *fanotify.Fanotifier
+	accessedFiles      []string
+	persistentPath     string
+	WaitSignal         bool
+	WaitLine           string
+	WaitTime           time.Duration
+	fanotifierClosed   bool
+	fanotifierClosedMu sync.Mutex
+}
+
+func GenerateFanotifyOpts(cmd *cobra.Command, flagT bool) (*Context, error) {
```

Comment:
Can we create a `FanotifyOptions` struct here and construct it from the `cmd` in `cmd/nerdctl/fanotify.go` and pass it to `pkg`, instead of passing a cobra `cmd` to `pkg` side.

Reply:
Also the flag validation logic should be on `cmd` side as well. 

In general, we should read cobra flags and do necessary validation on `cmd` side, (and also construct the `Option` struct from those flags. Then the main logic should be in `pkg` side, using the passed `Option` Struct.

Reply:
Yes, thanks for your explanation. I will refactor the code.